# ================================================================================================
# PROJECT 2 - CREDIT RISK ANALYSIS
# ANALYST - Rajeev Kumar 
# Domain    — BFSI — Credit Risk
# ================================================================================================

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 
import warnings

warnings.filterwarnings('ignore')
pd.set_option('display.float_format', lambda x: '%.2f' % x)
pd.set_option('display.max_columns', 50)

print("✅ All Libraries Imported Successfully!")

✅ All Libraries Imported Successfully!


In [2]:
df_applicants = pd.read_csv(r"C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\applicant_details.csv")
df_loans = pd.read_csv(r"C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\loan_details.csv")
df_repayments = pd.read_csv(r"C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\repayment_history.csv")

print("✅ All 3 Tables Loaded Successfully!")
print(f"\nApplicant Details  : {df_applicants.shape[0]:,} rows x {df_applicants.shape[1]} columns")
print(f"Loan Details       : {df_loans.shape[0]:,} rows x {df_loans.shape[1]} columns")
print(f"Repayment History  : {df_repayments.shape[0]:,} rows x {df_repayments.shape[1]} columns")
print(f"\nTotal Records      : {df_applicants.shape[0] + df_loans.shape[0] + df_repayments.shape[0]:,}")

✅ All 3 Tables Loaded Successfully!

Applicant Details  : 20,000 rows x 14 columns
Loan Details       : 30,000 rows x 12 columns
Repayment History  : 352,500 rows x 11 columns

Total Records      : 402,500


In [3]:
# First Look at All 3 Tables
print("=" * 60)
print("TABLE 1 — APPLICANT DETAILS")
print("=" * 60)
print(df_applicants.head(3))
print(f"\nColumns: {list(df_applicants.columns)}")

print("\n" + "=" * 60)
print("TABLE 2 — LOAN DETAILS")
print("=" * 60)
print(df_loans.head(3))
print(f"\nColumns: {list(df_loans.columns)}")

print("\n" + "=" * 60)
print("TABLE 3 — REPAYMENT HISTORY")
print("=" * 60)
print(df_repayments.head(3))
print(f"\nColumns: {list(df_repayments.columns)}")

TABLE 1 — APPLICANT DETAILS
  Applicant_ID Applicant_Name  Age  Gender         City Region  \
0     APP00001  Kavita Mishra   42    Male    Faridabad  North   
1     APP00002    manoj joshi   53  Female  Bhubaneswar   East   
2     APP00003     Nitin Nair   33  Female      Kolkata   East   

  Employment_Type      Education Marital_Status  Annual_Income  Credit_Score  \
0        Salaried  Post Graduate         Single      695112.00        698.00   
1   Self-Employed      Doctorate       Divorced     1186969.00        761.00   
2        Salaried  Post Graduate       Divorced      817660.00        648.00   

   Dependents  Years_At_Job Own_House  
0           1             0       Yes  
1           0             4       Yes  
2           1             1       Yes  

Columns: ['Applicant_ID', 'Applicant_Name', 'Age', 'Gender', 'City', 'Region', 'Employment_Type', 'Education', 'Marital_Status', 'Annual_Income', 'Credit_Score', 'Dependents', 'Years_At_Job', 'Own_House']

TABLE 2 — LOAN DETA

# Data Types Check  All 3 Tables

In [4]:
print(df_applicants.dtypes)

Applicant_ID        object
Applicant_Name      object
Age                  int64
Gender              object
City                object
Region              object
Employment_Type     object
Education           object
Marital_Status      object
Annual_Income      float64
Credit_Score       float64
Dependents           int64
Years_At_Job         int64
Own_House           object
dtype: object


In [5]:
print(df_loans.dtypes)

Loan_ID                object
Applicant_ID           object
Loan_Type              object
Loan_Amount           float64
Interest_Rate         float64
Loan_Tenure_Months      int64
Loan_Purpose           object
Loan_Grade             object
EMI_Amount            float64
Issue_Date             object
Debt_To_Income_%      float64
Loan_Status            object
dtype: object


In [6]:
print(df_repayments.dtypes)

Repayment_ID       object
Loan_ID            object
Applicant_ID       object
EMI_Month           int64
Due_Date           object
Paid_Date          object
EMI_Amount        float64
Amount_Paid       float64
DPD_Days            int64
Payment_Status     object
Default_Flag        int64
dtype: object


## Observation — Data Types

Table 1 — Applicant Details
1. Annual_Income float64 — should be int — missing values cause
2. Credit_Score float64 — should be int — missing values cause

Table 2 — Loan Details
1. Issue_Date object — should be datetime
2. Loan_Amount float64 — has missing values

Table 3 — Repayment History
1. Due_Date object — should be datetime
2. Paid_Date object — should be datetime
3. Default_Flag int64 — correct ✅
4. DPD_Days int64 — correct ✅

Action Plan:
- Convert date columns to datetime
- Fix float to int after handling missing values

# Missing Values Check

In [7]:
print('Applicants table missing details ')
print(df_applicants.isnull().sum())
print(f"\nTotal Missing : {df_applicants.isnull().sum().sum()}")
print(f"Missing %     : {round(df_applicants.isnull().sum().sum() / (df_applicants.shape[0] * df_applicants.shape[1]) * 100, 2)}%")


Applicants table missing details 
Applicant_ID         0
Applicant_Name       0
Age                  0
Gender               0
City                 0
Region               0
Employment_Type      0
Education          407
Marital_Status       0
Annual_Income      623
Credit_Score       404
Dependents           0
Years_At_Job         0
Own_House            0
dtype: int64

Total Missing : 1434
Missing %     : 0.51%


In [8]:
print("\n" + "-" * 40)
print("TABLE 2 — LOAN DETAILS — Missing Values")
print(df_loans.isnull().sum())
print(f"\nTotal Missing : {df_loans.isnull().sum().sum()}")
print(f"Missing %     : {round(df_loans.isnull().sum().sum() / (df_loans.shape[0] * df_loans.shape[1]) * 100, 2)}%")

print("\n" + "-" * 40)
print("TABLE 3 — REPAYMENT HISTORY — Missing Values")
print(df_repayments.isnull().sum())
print(f"\nTotal Missing : {df_repayments.isnull().sum().sum()}")
print(f"Missing %     : {round(df_repayments.isnull().sum().sum() / (df_repayments.shape[0] * df_repayments.shape[1]) * 100, 2)}%")


----------------------------------------
TABLE 2 — LOAN DETAILS — Missing Values
Loan_ID                 0
Applicant_ID            0
Loan_Type               0
Loan_Amount           593
Interest_Rate         600
Loan_Tenure_Months      0
Loan_Purpose            0
Loan_Grade              0
EMI_Amount              0
Issue_Date              0
Debt_To_Income_%        0
Loan_Status             0
dtype: int64

Total Missing : 1193
Missing %     : 0.33%

----------------------------------------
TABLE 3 — REPAYMENT HISTORY — Missing Values
Repayment_ID          0
Loan_ID               0
Applicant_ID          0
EMI_Month             0
Due_Date              0
Paid_Date         27518
EMI_Amount            0
Amount_Paid        6976
DPD_Days              0
Payment_Status     7068
Default_Flag          0
dtype: int64

Total Missing : 41562
Missing %     : 1.07%


## Observation — Missing Values

Table 1 — Applicant Details
1. Annual_Income — 623 missing (3.1%) — data entry gap
2. Credit_Score — 404 missing (2.0%) — data entry gap
3. Education — 407 missing (2.0%) — data entry gap
4. Total missing — 1,434 — only 0.51% — very low

Table 2 — Loan Details
1. Loan_Amount — 593 missing (1.9%) — data entry gap
2. Interest_Rate — 600 missing (2.0%) — data entry gap
3. Total missing — 1,193 — only 0.33% — very low

Table 3 — Repayment History
1. Paid_Date — 27,518 missing — EXPECTED!
   Defaulted accounts have no payment date — not a problem
2. Payment_Status — 7,068 missing — data entry gap
3. Amount_Paid — 6,976 missing — data entry gap
4. Total missing — 41,562 — 1.07% — acceptable

Overall data quality is good — missing values under 3%
Paid_Date missing is business logic — not data quality issue

# Duplicate Check

In [9]:
print(f"Table 1 — Applicant Details  : {df_applicants.duplicated().sum()} duplicates")
print(f"Table 2 — Loan Details       : {df_loans.duplicated().sum()} duplicates")
print(f"Table 3 — Repayment History  : {df_repayments.duplicated().sum()} duplicates")


Table 1 — Applicant Details  : 0 duplicates
Table 2 — Loan Details       : 0 duplicates
Table 3 — Repayment History  : 0 duplicates


In [10]:
print("\nDUPLICATE IDs CHECK")
print("-" * 40)
print(f"Duplicate Applicant_IDs  : {df_applicants['Applicant_ID'].duplicated().sum()}")
print(f"Duplicate Loan_IDs       : {df_loans['Loan_ID'].duplicated().sum()}")
print(f"Duplicate Repayment_IDs  : {df_repayments['Repayment_ID'].duplicated().sum()}")


DUPLICATE IDs CHECK
----------------------------------------
Duplicate Applicant_IDs  : 0
Duplicate Loan_IDs       : 0
Duplicate Repayment_IDs  : 0


In [11]:
print("\nAPPLICANTS WITH MULTIPLE LOANS")
print("-" * 40)

multi_loans = df_loans.groupby('Applicant_ID')['Loan_ID'].count()
print(f"Applicants with 1 loan   : {(multi_loans == 1).sum():,}")
print(f"Applicants with 2 loans  : {(multi_loans == 2).sum():,}")
print(f"Applicants with 3+ loans : {(multi_loans >= 3).sum():,}")
print(f"Max loans by one person  : {multi_loans.max()}")


APPLICANTS WITH MULTIPLE LOANS
----------------------------------------
Applicants with 1 loan   : 6,707
Applicants with 2 loans  : 4,995
Applicants with 3+ loans : 3,848
Max loans by one person  : 10


## Observation — Duplicates & Multiple Loans

1. No duplicate rows in any table — data is clean
2. No duplicate IDs — all primary keys are unique

3. Multiple Loans Analysis
   - 43.5% applicants have 1 loan
   - 32.4% applicants have 2 loans
   - 24.9% applicants have 3+ loans
   - One person has maximum 10 loans!

4. Business Insight
   - 56.5% customers have multiple loans — cross selling is high
   - Multiple loans = higher debt burden = higher default risk
   - Max 10 loans by one person is a red flag for credit risk
   - This is why we need Debt_To_Income_% as a key metric

# Unique Values Check — Categorical Columns

## TABLE 1 — APPLICANT DETAILS

In [12]:
cat_cols_app = ['Gender', 'Region', 'Employment_Type', 'Education', 'Marital_Status', 'Own_House']

for col in cat_cols_app: 
    print(f'\n{col} - {df_applicants[col].nunique()} unique values')
    print(df_applicants[col].value_counts())


Gender - 5 unique values
Gender
Male      11742
Female     7867
F           138
M           130
MALE        123
Name: count, dtype: int64

Region - 5 unique values
Region
Central    4088
East       4043
South      3974
North      3957
West       3938
Name: count, dtype: int64

Employment_Type - 5 unique values
Employment_Type
Business Owner         4068
Salaried               4025
Freelancer             3995
Self-Employed          3981
Government Employee    3931
Name: count, dtype: int64

Education - 4 unique values
Education
Doctorate        4963
Graduate         4891
Post Graduate    4887
High School      4852
Name: count, dtype: int64

Marital_Status - 4 unique values
Marital_Status
Single      5059
Widowed     5052
Divorced    4973
Married     4916
Name: count, dtype: int64

Own_House - 2 unique values
Own_House
No     11986
Yes     8014
Name: count, dtype: int64


## TABLE 2 — LOAN DETAILS

In [13]:
cat_cols_loan = ['Loan_Type', 'Loan_Grade', 'Loan_Purpose', 'Loan_Status']

for col in cat_cols_loan: 
    print(f"\n{col} — {df_loans[col].nunique()} unique values")
    print(df_loans[col].value_counts())


Loan_Type — 6 unique values
Loan_Type
Personal Loan     5124
Home Loan         5041
Auto Loan         5007
Education Loan    5005
Gold Loan         5000
Business Loan     4823
Name: count, dtype: int64

Loan_Grade — 6 unique values
Loan_Grade
D    10271
E     7320
C     6250
F     3142
B     2339
A      678
Name: count, dtype: int64

Loan_Purpose — 9 unique values
Loan_Purpose
Medical Emergency     3477
Debt Consolidation    3341
Home Purchase         3337
Wedding               3335
Business Expansion    3328
Home Renovation       3307
Investment            3302
Education             3296
Vehicle Purchase      3277
Name: count, dtype: int64

Loan_Status — 2 unique values
Loan_Status
Active    23908
Closed     6092
Name: count, dtype: int64


## TABLE 3 — REPAYMENT HISTORY

In [14]:
cat_cols_rep = ['Payment_Status']
for col in cat_cols_rep:
    print(f"\n{col} — {df_repayments[col].nunique()} unique values")
    print(df_repayments[col].value_counts())


Payment_Status — 3 unique values
Payment_Status
On Time      272849
Late          45606
Defaulted     26977
Name: count, dtype: int64


## Unique Values — What I Found

**Applicant Details**
- Gender column has 5 unique values but it should only have 2
  - Male, M and MALE are all the same thing — someone was inconsistent while entering data
  - Same issue with Female and F
  - I will fix this by mapping all variations to Male/Female
- Region, Employment Type, Education, Marital Status and Own House all look clean — no issues found

**Loan Details**
- The Loan Grade distribution caught my attention
  - Grade D has the most loans — 10,271 (34%) — this is medium-high risk
  - Grade A has the fewest — only 678 loans (2.3%) — very few low risk borrowers
  - This means the portfolio is heavily skewed towards risky borrowers which explains the high default rate we saw earlier
- Loan Type and Loan Purpose look clean
- 79.7% loans are still Active and 20.3% are Closed — makes sense

**Repayment History**
- Payment Status breakdown is quite revealing
  - 77.4% payments are On Time — this is the good news
  - 12.9% are Late — these customers are struggling but still paying
  - 7.7% are Defaulted — these are the problem accounts
  - Overall 20.6% payments are either late or defaulted which is a serious concern for the

# Basic Statistics — Numerical Columns

In [15]:
print(df_applicants[['Age', 'Annual_Income', 'Credit_Score', 
                      'Dependents', 'Years_At_Job']].describe().round(2))

           Age  Annual_Income  Credit_Score  Dependents  Years_At_Job
count 20000.00       19377.00      19596.00    20000.00      20000.00
mean     37.80      805606.01        672.69        1.64          3.49
std       9.55      403462.44         59.92        1.32          3.97
min      21.00      150000.00        455.00        0.00          0.00
25%      31.00      523830.00        632.00        1.00          1.00
50%      38.00      719793.00        671.00        2.00          2.00
75%      44.00      993421.00        712.00        3.00          5.00
max      65.00     2820566.00        900.00        4.00         36.00


In [16]:
print(df_loans[['Loan_Amount', 'Interest_Rate', 'Loan_Tenure_Months',
                'EMI_Amount', 'Debt_To_Income_%']].describe().round(2))

       Loan_Amount  Interest_Rate  Loan_Tenure_Months  EMI_Amount  \
count     29407.00       29400.00            30000.00    30000.00   
mean     703268.75          16.30               71.10    20183.37   
std      816457.67           1.31               85.48    18293.36   
min       20000.00          11.08                6.00      977.92   
25%      204000.00          15.39               24.00     8308.72   
50%      415000.00          16.32               36.00    14692.53   
75%      860000.00          17.23               60.00    26343.30   
max     4999000.00          20.84              360.00   181939.05   

       Debt_To_Income_%  
count          30000.00  
mean              33.85  
std               28.10  
min                0.73  
25%               14.63  
50%               26.32  
75%               45.26  
max              218.57  


In [17]:
print(df_repayments[['EMI_Amount', 'Amount_Paid', 
                     'DPD_Days', 'Default_Flag']].describe().round(2))

       EMI_Amount  Amount_Paid  DPD_Days  Default_Flag
count   352500.00    345524.00 352500.00     352500.00
mean     20116.29     17879.39     12.67          0.12
std      18380.17     18187.45     36.79          0.32
min        977.92         0.00      0.00          0.00
25%       8258.91      6233.96      0.00          0.00
50%      14582.84     12709.26      0.00          0.00
75%      26150.17     23936.22      0.00          0.00
max     181939.05    181939.05    180.00          1.00


## Statistics — What I Found

**Applicant Details**
- Average age is 38 years which makes sense for loan applicants
- Average income is 8 Lakhs but max is 28 Lakhs — shows wide gap between rich and poor applicants
- Credit score average is 673 — decent but not great — many applicants are borderline risky

**Loan Details**
- Average loan amount is 7 Lakhs with very high variation — some took 20K some took 50 Lakhs
- Interest rates range from 11% to 21% — realistic for Indian BFSI market
- DTI% average is 33.85% — already above safe limit of 30%
- Most alarming — someone has DTI of 218%! They are paying more than twice their income as EMI — this is extreme default risk

**Repayment History**
- Default rate is 12% — industry average is 5-8% — our portfolio is clearly high risk
- Average amount paid (17,879) is less than average EMI (20,116)
- This means we are losing around 2,237 per EMI record — significant revenue loss
- Maximum DPD is 180 days — some accounts are 6 months overdue

In [18]:
df_applicants_original = df_applicants.copy()
df_loans_original      = df_loans.copy()
df_repayments_original = df_repayments.copy()

In [19]:
print(f"Applicants  : {df_applicants_original.shape}")
print(f"Loans       : {df_loans_original.shape}")
print(f"Repayments  : {df_repayments_original.shape}")

Applicants  : (20000, 14)
Loans       : (30000, 12)
Repayments  : (352500, 11)


## Data Cleaning

Before touching any data I always take a backup.
This way if something goes wrong during cleaning
I can always go back to the original data and start fresh.

Problems I need to fix:
- Gender column has inconsistent formats (Male, M, MALE)
- Missing values in Income, Credit Score, Education
- Missing values in Loan Amount and Interest Rate
- Date columns stored as text — need datetime conversion
- Float columns that should be int
- Extreme outlier in DTI% (218%)

# Fix Gender Column — Inconsistent Formats

In [20]:
print("BEFORE FIXING GENDER")
print(df_applicants['Gender'].value_counts())

BEFORE FIXING GENDER
Gender
Male      11742
Female     7867
F           138
M           130
MALE        123
Name: count, dtype: int64


In [21]:
gender_maping = {
    'Male':'Male', 
    'Female':'Female', 
    'M': 'Male', 
    'F': 'Female', 
    'MALE': 'Male', 
    'FEMALE':'Female'
}

In [22]:
df_applicants['Gender'] = df_applicants['Gender'].map(gender_maping)

In [23]:
print("\nAFTER FIXING GENDER")
print(df_applicants['Gender'].value_counts())
print(f"\nUnique values now : {df_applicants['Gender'].nunique()}")


AFTER FIXING GENDER
Gender
Male      11995
Female     8005
Name: count, dtype: int64

Unique values now : 2


# Fix Missing Values 

In [24]:
print("BEFORE FIXING MISSING VALUES") 

print(f'Applicants : {df_applicants.isnull().sum().sum()} missing')
print(f'Loans : {df_loans.isnull().sum().sum()} missing')
print(f'Repayments : {df_repayments.isnull().sum().sum()} missing')

BEFORE FIXING MISSING VALUES
Applicants : 1434 missing
Loans : 1193 missing
Repayments : 41562 missing


In [25]:
# Annual_Income :  fill with median (income is skewed median better than mean) 


df_applicants['Annual_Income'] = df_applicants['Annual_Income'].fillna(
    df_applicants['Annual_Income'].median()
)

In [26]:
# Credit_Score — fill with median


df_applicants['Credit_Score'] = df_applicants['Credit_Score'].fillna(
    df_applicants['Credit_Score'].median()
)

In [27]:
# Education — fill with mode (most common value)

df_applicants['Education'] = df_applicants['Education'].fillna(
    df_applicants['Education'].mode()[0]
)

In [28]:
df_applicants.isnull().sum()


Applicant_ID       0
Applicant_Name     0
Age                0
Gender             0
City               0
Region             0
Employment_Type    0
Education          0
Marital_Status     0
Annual_Income      0
Credit_Score       0
Dependents         0
Years_At_Job       0
Own_House          0
dtype: int64

# ── Table 2 — Loan Details ────────────────────────────────────

In [29]:
df_loans.isnull().sum()

Loan_ID                 0
Applicant_ID            0
Loan_Type               0
Loan_Amount           593
Interest_Rate         600
Loan_Tenure_Months      0
Loan_Purpose            0
Loan_Grade              0
EMI_Amount              0
Issue_Date              0
Debt_To_Income_%        0
Loan_Status             0
dtype: int64

In [30]:
df_loans['Loan_Amount'] = df_loans['Loan_Amount'].fillna(
    df_loans['Loan_Amount'].median())

In [31]:
df_loans['Interest_Rate'] = df_loans['Interest_Rate'].fillna(
    df_loans['Interest_Rate'].median())

In [32]:
df_loans.isnull().sum()

Loan_ID               0
Applicant_ID          0
Loan_Type             0
Loan_Amount           0
Interest_Rate         0
Loan_Tenure_Months    0
Loan_Purpose          0
Loan_Grade            0
EMI_Amount            0
Issue_Date            0
Debt_To_Income_%      0
Loan_Status           0
dtype: int64

# ── Table 3 — Repayment History ───────────────────────────────

In [33]:
df_repayments['Amount_Paid'] = df_repayments['Amount_Paid'].fillna(0)

In [34]:
df_repayments['Payment_Status'] = df_repayments['Payment_Status'].fillna('Unknown')


In [35]:
print("\nAFTER FIXING MISSING VALUES")
print(f"Applicants  : {df_applicants.isnull().sum().sum()} missing")
print(f"Loans       : {df_loans.isnull().sum().sum()} missing")
print(f"Repayments  : {df_repayments.isnull().sum().sum()} missing")


AFTER FIXING MISSING VALUES
Applicants  : 0 missing
Loans       : 0 missing
Repayments  : 27518 missing


In [36]:
df_repayments.isnull().sum()

Repayment_ID          0
Loan_ID               0
Applicant_ID          0
EMI_Month             0
Due_Date              0
Paid_Date         27518
EMI_Amount            0
Amount_Paid           0
DPD_Days              0
Payment_Status        0
Default_Flag          0
dtype: int64

# Fix Date Columns — Convert Text to Datetime

In [37]:
print("BEFORE FIXING DATES")

print(f"Issue_Date type  : {df_loans['Issue_Date'].dtype}")
print(f"Due_Date type    : {df_repayments['Due_Date'].dtype}")
print(f"Paid_Date type   : {df_repayments['Paid_Date'].dtype}")

BEFORE FIXING DATES
Issue_Date type  : object
Due_Date type    : object
Paid_Date type   : object


In [38]:
df_loans['Issue_Date'] = pd.to_datetime(df_loans['Issue_Date'])
df_repayments['Due_Date'] = pd.to_datetime(df_repayments['Due_Date'])
df_repayments['Paid_Date'] = pd.to_datetime(df_repayments['Paid_Date'])

In [39]:
print("After FIXING DATES")

print(f"Issue_Date type  : {df_loans['Issue_Date'].dtype}")
print(f"Due_Date type    : {df_repayments['Due_Date'].dtype}")
print(f"Paid_Date type   : {df_repayments['Paid_Date'].dtype}")

After FIXING DATES
Issue_Date type  : datetime64[ns]
Due_Date type    : datetime64[ns]
Paid_Date type   : datetime64[ns]


In [40]:
df_loans['Issue_Year'] = df_loans['Issue_Date'].dt.year
df_loans['Issue_Month'] = df_loans['Issue_Date'].dt.month
df_loans['Issue_Quarter'] = df_loans['Issue_Date'].dt.quarter

In [41]:
df_loans.head(3)

,Loan_ID,Applicant_ID,Loan_Type,Loan_Amount,Interest_Rate,Loan_Tenure_Months,Loan_Purpose,Loan_Grade,EMI_Amount,Issue_Date,Debt_To_Income_%,Loan_Status,Issue_Year,Issue_Month,Issue_Quarter
0,LN000001,APP16888,Personal Loan,77000.00,16.19,60,Wedding,D,1880.27,2024-03-09,2.48,Active,2024,3,1
1,LN000002,APP19301,Auto Loan,787000.00,14.99,24,Debt Consolidation,D,38155.25,2025-01-23,48.66,Active,2025,1,1
2,LN000003,APP05290,Business Loan,1034000.00,17.85,24,Home Renovation,E,51546.61,2024-06-12,106.37,Active,2024,6,2


# Fix Data Types

In [42]:
print("BEFORE FIXING DATA TYPES")
print(f"Annual_Income : {df_applicants['Annual_Income'].dtype}")
print(f"Credit_Score  : {df_applicants['Credit_Score'].dtype}")

BEFORE FIXING DATA TYPES
Annual_Income : float64
Credit_Score  : float64


In [43]:
df_applicants['Annual_Income'] = df_applicants['Annual_Income'].astype(int)
df_applicants['Credit_Score'] = df_applicants['Credit_Score'].astype(int)

In [44]:
print("\nAFTER FIXING DATA TYPES")
print(f"Annual_Income : {df_applicants['Annual_Income'].dtype}")
print(f"Credit_Score  : {df_applicants['Credit_Score'].dtype}")


AFTER FIXING DATA TYPES
Annual_Income : int64
Credit_Score  : int64


## Cleaning Step 4 — Fixed Data Types

Annual_Income and Credit_Score were stored as float64
because they had missing values — pandas automatically
converts int columns to float when there are NaN values.

Now that I have filled all missing values I can safely
convert them back to int. Income and credit scores are
always whole numbers so int makes more sense here.

# Feature Engineering — Creating New Useful Columns

## Applicants Table

### 1. Credit Score Category

In [45]:
df_applicants['Credit_Score_Category'] = pd.cut(
    df_applicants['Credit_Score'], 
    bins=[0,580,670,740,800,900], 
    labels= ['Poor','Fair','Good','Very Good','Excellent']
)

### 2. Age Group

In [46]:
df_applicants['Age_Group'] = pd.cut(
    df_applicants['Age'], 
    bins=[20,30,40,50,65], 
    labels=['21-30','31-40','41-50','51-65']
)

### 3. Income Category

In [47]:
df_applicants['Income_Category'] = pd.cut(
    df_applicants['Annual_Income'],
    bins=[0, 300000, 600000, 1000000, 2000000, 99999999],
    labels=['Low', 'Lower Middle', 'Middle', 'Upper Middle', 'High']
)


## Loans Table

### 1. DTI Risk Category

In [48]:
df_loans['DTI_Risk'] = pd.cut(
    df_loans['Debt_To_Income_%'],
    bins=[0, 20, 35, 50, 999],
    labels=['Low Risk', 'Moderate Risk', 'High Risk', 'Very High Risk']
)


### 2. Loan Size Category

In [49]:
df_loans['Loan_Size'] = pd.cut(
    df_loans['Loan_Amount'],
    bins=[0, 100000, 500000, 1000000, 99999999],
    labels=['Small', 'Medium', 'Large', 'Very Large']
)


In [50]:
print("New columns added to Applicants:")
print(df_applicants[['Credit_Score', 'Credit_Score_Category',
                      'Age', 'Age_Group',
                      'Annual_Income', 'Income_Category']].head(5))

New columns added to Applicants:
   Credit_Score Credit_Score_Category  Age Age_Group  Annual_Income  \
0           698                  Good   42     41-50         695112   
1           761             Very Good   53     51-65        1186969   
2           648                  Fair   33     31-40         817660   
3           680                  Good   33     31-40         332988   
4           605                  Fair   27     21-30         601850   

  Income_Category  
0          Middle  
1    Upper Middle  
2          Middle  
3    Lower Middle  
4          Middle  


In [51]:
print("New columns added to Loans:")
print(df_loans[['Debt_To_Income_%', 'DTI_Risk',
                'Loan_Amount', 'Loan_Size']].head(5))

New columns added to Loans:
   Debt_To_Income_%        DTI_Risk  Loan_Amount   Loan_Size
0              2.48        Low Risk     77000.00       Small
1             48.66       High Risk    787000.00       Large
2            106.37  Very High Risk   1034000.00  Very Large
3             35.17       High Risk     98000.00       Small
4             28.60   Moderate Risk    707000.00       Large


# Merge All 3 Tables

## Applicants + Loans Table

In [52]:
df_app_loans = pd.merge(
    df_applicants, 
    df_loans, 
    on='Applicant_ID', 
    how='left'
)

In [53]:
print(f"After Merge 1 (Applicants + Loans) : {df_app_loans.shape}")


After Merge 1 (Applicants + Loans) : (34450, 33)


## App_loans + Repayments 

In [54]:
df_merged = pd.merge(
    df_app_loans,
    df_repayments,
    on='Loan_ID', 
    how='left'
)

In [55]:
print(f"After Merge 2 (+ Repayments)       : {df_merged.shape}")

After Merge 2 (+ Repayments)       : (356950, 43)


In [56]:
print(f"\nTotal columns in merged table : {df_merged.shape[1]}")
print(f"\nSample columns:")
print(list(df_merged.columns))


Total columns in merged table : 43

Sample columns:
['Applicant_ID_x', 'Applicant_Name', 'Age', 'Gender', 'City', 'Region', 'Employment_Type', 'Education', 'Marital_Status', 'Annual_Income', 'Credit_Score', 'Dependents', 'Years_At_Job', 'Own_House', 'Credit_Score_Category', 'Age_Group', 'Income_Category', 'Loan_ID', 'Loan_Type', 'Loan_Amount', 'Interest_Rate', 'Loan_Tenure_Months', 'Loan_Purpose', 'Loan_Grade', 'EMI_Amount_x', 'Issue_Date', 'Debt_To_Income_%', 'Loan_Status', 'Issue_Year', 'Issue_Month', 'Issue_Quarter', 'DTI_Risk', 'Loan_Size', 'Repayment_ID', 'Applicant_ID_y', 'EMI_Month', 'Due_Date', 'Paid_Date', 'EMI_Amount_y', 'Amount_Paid', 'DPD_Days', 'Payment_Status', 'Default_Flag']


# Fix Duplicate Columns After Merge

In [57]:
print("Columns with _x or _y suffix:")
dup_cols = [col for col in df_merged.columns if '_x' in col or '_y' in col]
print(dup_cols)

Columns with _x or _y suffix:
['Applicant_ID_x', 'EMI_Amount_x', 'Applicant_ID_y', 'EMI_Amount_y']


In [58]:
df_merged = df_merged.drop(columns=['Applicant_ID_y', 'EMI_Amount_x'])

In [59]:
df_merged = df_merged.rename(columns={
    'Applicant_ID_x':'Applicant_ID', 
    'EMI_Amount_y' : 'EMI_Amount'
})

In [60]:
print("\nAfter fixing duplicate columns:")
print(f"Total columns : {df_merged.shape[1]}")
print(f"Columns : {list(df_merged.columns)}")


After fixing duplicate columns:
Total columns : 41
Columns : ['Applicant_ID', 'Applicant_Name', 'Age', 'Gender', 'City', 'Region', 'Employment_Type', 'Education', 'Marital_Status', 'Annual_Income', 'Credit_Score', 'Dependents', 'Years_At_Job', 'Own_House', 'Credit_Score_Category', 'Age_Group', 'Income_Category', 'Loan_ID', 'Loan_Type', 'Loan_Amount', 'Interest_Rate', 'Loan_Tenure_Months', 'Loan_Purpose', 'Loan_Grade', 'Issue_Date', 'Debt_To_Income_%', 'Loan_Status', 'Issue_Year', 'Issue_Month', 'Issue_Quarter', 'DTI_Risk', 'Loan_Size', 'Repayment_ID', 'EMI_Month', 'Due_Date', 'Paid_Date', 'EMI_Amount', 'Amount_Paid', 'DPD_Days', 'Payment_Status', 'Default_Flag']


In [61]:
df_merged.sample(1)

,Applicant_ID,Applicant_Name,Age,Gender,City,Region,Employment_Type,Education,Marital_Status,Annual_Income,Credit_Score,Dependents,Years_At_Job,Own_House,Credit_Score_Category,Age_Group,Income_Category,Loan_ID,Loan_Type,Loan_Amount,Interest_Rate,Loan_Tenure_Months,Loan_Purpose,Loan_Grade,Issue_Date,Debt_To_Income_%,Loan_Status,Issue_Year,Issue_Month,Issue_Quarter,DTI_Risk,Loan_Size,Repayment_ID,EMI_Month,Due_Date,Paid_Date,EMI_Amount,Amount_Paid,DPD_Days,Payment_Status,Default_Flag
298699,APP16749,Divya Kumar,43,Female,Ahmedabad,Central,Freelancer,Post Graduate,Married,609403,746,2,0,Yes,Very Good,41-50,Middle,LN021810,Personal Loan,101000.00,14.44,60.00,Debt Consolidation,C,2023-11-24,4.67,Active,2023.00,11.00,4.00,Low Risk,Medium,REP0256156,10.00,2024-09-19,2024-09-19,2373.20,2373.20,0.00,On Time,0.00


## Fixed Duplicate Columns After Merge

When merging tables that have columns with same names
pandas automatically adds _x and _y suffixes to
distinguish them.

I found 4 such columns:
- Applicant_ID_x and Applicant_ID_y — same column
  dropped _y version and renamed _x to Applicant_ID
- EMI_Amount_x (from loans) and EMI_Amount_y (from repayments)
  kept repayments version and renamed it to EMI_Amount

Now all column names are clean without any suffixes.

In [68]:

df_merged.isnull().sum()

Applicant_ID                 0
Applicant_Name               0
Age                          0
Gender                       0
City                         0
Region                       0
Employment_Type              0
Education                    0
Marital_Status               0
Annual_Income                0
Credit_Score                 0
Dependents                   0
Years_At_Job                 0
Own_House                    0
Credit_Score_Category        0
Age_Group                    0
Income_Category              0
Loan_ID                   4450
Loan_Type                 4450
Loan_Amount               4450
Interest_Rate             4450
Loan_Tenure_Months        4450
Loan_Purpose              4450
Loan_Grade                4450
Issue_Date                4450
Debt_To_Income_%          4450
Loan_Status               4450
Issue_Year                4450
Issue_Month               4450
Issue_Quarter             4450
DTI_Risk                  4450
Loan_Size                 4450
Repaymen

In [69]:
df_merged.shape

(356950, 41)

In [72]:
print(f"Applicants table missing : {df_applicants.isnull().sum().sum()}")
print(f"Applicants table missing : {df_loans.isnull().sum().sum()}")
print(f"Applicants table missing : {df_repayments.isnull().sum().sum()}")

Applicants table missing : 0
Applicants table missing : 0
Applicants table missing : 27518


In [74]:
print(df_merged.isnull().sum().sum())

134318


In [75]:
# Check which columns have missing values in merged table
missing = df_merged.isnull().sum()
missing = missing[missing > 0]
print("Columns with missing values in merged table:")
print(missing)
print(f"\nTotal : {missing.sum()}")

Columns with missing values in merged table:
Loan_ID                4450
Loan_Type              4450
Loan_Amount            4450
Interest_Rate          4450
Loan_Tenure_Months     4450
Loan_Purpose           4450
Loan_Grade             4450
Issue_Date             4450
Debt_To_Income_%       4450
Loan_Status            4450
Issue_Year             4450
Issue_Month            4450
Issue_Quarter          4450
DTI_Risk               4450
Loan_Size              4450
Repayment_ID           4450
EMI_Month              4450
Due_Date               4450
Paid_Date             31968
EMI_Amount             4450
Amount_Paid            4450
DPD_Days               4450
Payment_Status         4450
Default_Flag           4450
dtype: int64

Total : 134318


In [76]:
print(f"Before dropping : {df_merged.shape[0]:,} rows")

Before dropping : 356,950 rows


In [77]:
df_merged = df_merged.dropna(subset=['Loan_ID'])

In [78]:
print(f"After dropping  : {df_merged.shape[0]:,} rows")
print(f"Rows removed    : {4450} — applicants with no loans")

After dropping  : 352,500 rows
Rows removed    : 4450 — applicants with no loans


In [79]:
missing = df_merged.isnull().sum()
missing = missing[missing > 0]
print(f"\nRemaining missing values:")
print(missing)


Remaining missing values:
Paid_Date    27518
dtype: int64


In [83]:
float_cols = ['Loan_Amount','Loan_Tenure_Months','Issue_Year','Issue_Quarter','Issue_Month','EMI_Month','DPD_Days','Default_Flag' ]

In [88]:
df_merged.dtypes

Applicant_ID                     object
Applicant_Name                   object
Age                               int64
Gender                           object
City                             object
Region                           object
Employment_Type                  object
Education                        object
Marital_Status                   object
Annual_Income                     int64
Credit_Score                      int64
Dependents                        int64
Years_At_Job                      int64
Own_House                        object
Credit_Score_Category          category
Age_Group                      category
Income_Category                category
Loan_ID                          object
Loan_Type                        object
Loan_Amount                       int64
Interest_Rate                   float64
Loan_Tenure_Months                int64
Loan_Purpose                     object
Loan_Grade                       object
Issue_Date               datetime64[ns]


In [87]:
for col in float_cols: 
    print('************************')
    print(f'before fixing : {df_merged[col].dtypes}')
    df_merged[col] = df_merged[col].astype('int')
    print(f'before fixing : {df_merged[col].dtypes}')

************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64
************************
before fixing : float64
before fixing : int64


In [89]:
df_applicants.to_csv(r'C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\Clean_dataset\applicant_details_cleaned.csv', index=False)
df_loans.to_csv(r'C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\Clean_dataset\loan_details_cleaned.csv', index=False)
df_repayments.to_csv(r'C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\Clean_dataset\repayment_history_cleaned.csv', index=False)
df_merged.to_csv(r'C:\Data Analytics\Projects\END TO END\Credit_Risk_Analysis\Dataset\Clean_dataset\credit_risk_merged.csv', index=False)
